### PR TITLE
Tweak webservers init to offset connection spikes and calm git down

### PR DESCRIPTION
**Pull request by iHiD**
This primarily aims to fix the git issue.

For each repo, we need to run the following on Bastion.
```
cd /mnt/efs/repos/ruby
sudo chgrp -R exercism-git .
sudo chmod -R g+w .
sudo chmod g-w objects/pack/*
sudo find -type d -exec chmod g+s {} +
git config core.sharedRepository true
```

And also this to fix future repos:
```
cd /mnt/efs/repos/
sudo chgrp exercism-git .
sudo chmod g+s
```

We can safely do this before deploying this branch, but it won't fix the issues until we do.

### CODE REVIEW

File: app/models/git/repository.rb [hunk 1]
```diff
@@ -134,6 +134,7 @@ def rugged_repo
         cmd = [
           "git clone",
           "--bare",
+          "-c core.sharedRepository=true",
```
Comment (iHiD) [inline app/models/git/repository.rb:137]: This tells git to honour group not user. It's set in the `.git/config` of each repo. We need to manually update old ones but this should fix it for new ones.

File: bin/start_webserver [hunk 2]
```diff
@@ -21,4 +21,10 @@ FileUtils.chdir APP_ROOT do
 
   # Finally we get the actual server running
   system! "bundle exec bin/rails server -e production -b '0.0.0.0' -p 3000"
+
+  # As EFS files are owned by different users in different places
+  # we need to tell libgit to be less stressed about user conflicts.
+  Track.pluck(:slug).each do |slug|
+    system("git config --global --add safe.directory /mnt/efs/repos/#{slug}")
+  end
```
Comment (iHiD) [inline bin/start_webserver]: We have to set this on each new machine where any updates might get called. 

There are three:
- webservers
- sidekiq
- bastion

State:
- Bastion is manual
- This does webservers
- We need to add a new script for sidekiq.

File: docker/rails.Dockerfile [hunk 3]
```diff
@@ -37,10 +37,13 @@ COPY . ./
...
 # This compiles the assets
-# During deployment the assets are copied from this image and 
+# During deployment the assets are copied from this image and
 # uploaded into s3. The assets left on the machine are not actually
 # used leave the assets on here.
 RUN bundle exec rails r bin/monitor-manifest
 RUN bundle exec rails assets:precompile
 
+RUN groupadd -g 2222 exercism-git
+RUN usermod -a -G exercism-git root
```
Comment (iHiD) [inline docker/rails.Dockerfile:47]: This gets us using the same user group on all machines that might write to EFS.

File: lib/run_migrations_with_concurrent_guard.rb [hunk 4]
```diff
@@ -2,6 +2,10 @@
 # for any concurrent failures;
 
 begin
+  # Offset all the different rails c's against each other over 30secs
+  # Put it in this begin so it keeps on happening on each retry.
+  sleep(rand * 30)
```
Comment (iHiD) [inline lib/run_migrations_with_concurrent_guard.rb:7]: This is unrelated to this PR really, but it just stops us killing the db by having 20 new servers all try and run migrations simultaneously.

**Review by ErikSchierboom — APPROVED** [review]
LGTM

File: bin/start_sidekiq [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+def system!(*args)
+  system(*args) || abort("\n== Command #{args} failed ==")
+end
+
+FileUtils.chdir APP_ROOT do
+  # As EFS files are owned by different users in different places
+  # we need to tell libgit to be less stressed about user conflicts.
+  # Any instance of "fatal: detected dubious ownership in repository" 
+  # means that this hasn't worked properly.
+  system! "ls -1 /mnt/efs/repos | xargs -p -L1 git config --global --add safe.directory"
```
Comment (ErikSchierboom) [inline bin/start_sidekiq]:
```suggestion
  system! "ls -1 -d /mnt/efs/repos/* | xargs -L1 git config --global --add safe.directory"
```

File: bin/start_webserver [hunk 6]
```diff
@@ -14,6 +14,12 @@ FileUtils.chdir APP_ROOT do
   # We do this via this helper script to sleep and retry on conurrent failures
   system! 'bundle exec bin/rails runner lib/run_migrations_with_concurrent_guard.rb'
 
+  # As EFS files are owned by different users in different places
+  # we need to tell libgit to be less stressed about user conflicts.
+  # Any instance of "fatal: detected dubious ownership in repository" 
+  # means that this hasn't worked properly.
+  system! "ls -1 /mnt/efs/repos | xargs -p -L1 git config --global --add safe.directory"
```
Comment (ErikSchierboom) [inline bin/start_webserver]:
```suggestion
  system! "ls -1 -d /mnt/efs/repos/* | xargs -L1 git config --global --add safe.directory"
```

**Pull request merged**